src/Parser/DomainPart.php

```diff
@@ -386,16 +395,33 @@ protected function hasBrackets()
         return true;
     }
 
-    protected function checkLabelLength(array $prev)
+    /**
+     * @param string $label
+     */
+    protected function checkLabelLength($label)
     {
-        if ($this->lexer->token['type'] === EmailLexer::S_DOT &&
-            $prev['type'] === EmailLexer::GENERIC &&
-            strlen($prev['value']) > 63
-        ) {
+        if ($this->isLabelTooLong($label)) {
             $this->warnings[LabelTooLong::CODE] = new LabelTooLong();
         }
     }
 
+    /**
+     * @param string $label
+     * @return bool
+     */
+    private function isLabelTooLong($label)
+    {
+        if (preg_match('/[^\x00-\x7F]/', $label)) {
+            $label = idn_to_ascii($label, IDNA_DEFAULT, INTL_IDNA_VARIANT_UTS46, $idnaInfo);
+
+            if (!$label) {
```

Comment:
You could remove the `if` and directly return `return (bool) ($idnaInfo['errors'] | IDNA_ERROR_LABEL_TOO_LONG)` , label has passed the length test already so testing again in line 422 wouldn't be needed.